### DISCUSSION TITLE
INSERT Query with relationships set

### DISCUSSION THREAD

**Discussion by PaulMcF1987** [Budibase/budibase#5984] [Help]
I am trying to write an INSERT Query.

As per my set up, I have 1 page with 3 different displays... View, Edit, Clone...

View works as it should and edit works using the standard save Query that comes with the set up.

On the clone section, as it stands it does the same as the edit section and simply overwrites the data.

I wrote a query but it will not let me save it...

I have a `Customer` column which has been set up as a relationship with the `Customer Accounts` table.

When I write the query it tells me `Query Error: Unknown column 'Customer' in 'field list'`
If I then change this to the field that is shown in the table `fk_Account_RecordJob` and run the query, I get an error `Query Error: Incorrect integer value: '' for column 'fk_CustomerAccount_RecordJob' at row 1`

There is no Customer Field shown in the Database.

I have removed the 'fk_CustomerAccount_RecordJob' from the form as selecting the customer should automatically fill this field out based on the "relationship"

Can someone shed some light on this please? And the best practice if possible

**Comment by melohagan** [comment]
Hi @PaulMcF1987 

Could you share some screenshots of the query, and the schemas for the `Customer` and `Customer Accounts` tables please?

**Reply by PaulMcF1987** [reply]
@melohagan The tables were set up on Budibase, rather than using a pre existing table since I wasnt sure how to set a table up to use the relationship option

There is no screenshot as it wouldnt let me save it, essentially the query was:

`Insert into RecordJob(`
		`JobTitle, Customer, BookedBy, Datebooked, CustPrice, CustSurcharges, Service, JobStatus
)`
`VALUES (`
				`{{ Form.Fields.JobTitle }}, {{ Form.Fields.Customer }}, {{ Form.Fields.BookedBy }}, {{ Form.Fields.Datebooked }}, {{Form.Fields.CustPrice}}, {{Form.Fields.CustSurcharges}}, {{Form.Fields.Service}}, {{Form.Fields.JobStatus}}
);`

I also tried
`Insert into RecordJob(`
		`JobTitle, Customer, BookedBy, Datebooked, CustPrice, Service, JobStatus
)`
`VALUES (`
				`{{ Form.Fields.JobTitle }}, {{ Form.Fields.fk_CustomerAccount_RecordJob}}, {{ Form.Fields.BookedBy }}, {{ Form.Fields.Datebooked }}, {{Form.Fields.CustPrice}}, {{Form.Fields.Service}}, {{Form.Fields.JobStatus}}
);`


Is this what you wanted for the Record Job Schema (the Query Budibase used to create the table, pulled from HeidiSQL)
CREATE TABLE `RecordJob` (
	`id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
	`JobTitle` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	`fk_Account_RecordJob` INT(10) UNSIGNED NULL DEFAULT NULL,
	`BookedBy` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	`Datebooked` DATETIME NULL DEFAULT NULL,
	`CustPrice` FLOAT(8,2) NULL DEFAULT NULL,
	`Service` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	`JobStatus` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	PRIMARY KEY (`id`) USING BTREE,
	INDEX `recordjob_fk_account_recordjob_foreign` (`fk_Account_RecordJob`) USING BTREE,
	INDEX `recordjob_fk_driver_recordjob_foreign` (`fk_Driver_RecordJob`) USING BTREE,
	INDEX `recordjob_fk_vehicle_recordjob_foreign` (`fk_Vehicle_RecordJob`) USING BTREE,
	CONSTRAINT `recordjob_fk_account_recordjob_foreign` FOREIGN KEY (`fk_Account_RecordJob`) REFERENCES `Account` (`id`) ON UPDATE NO ACTION ON DELETE NO ACTION,
	CONSTRAINT `recordjob_fk_driver_recordjob_foreign` FOREIGN KEY (`fk_Driver_RecordJob`) REFERENCES `Driver` (`id`) ON UPDATE NO ACTION ON DELETE NO ACTION,
	CONSTRAINT `recordjob_fk_vehicle_recordjob_foreign` FOREIGN KEY (`fk_Vehicle_RecordJob`) REFERENCES `Vehicle` (`id`) ON UPDATE NO ACTION ON DELETE NO ACTION

(There are other rows in the table but not applicable to this query)

Customer Accounts (Same situation as above Query)
CREATE TABLE `Account` (
	`id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
	`Company` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	`AccountNo` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	`Address` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	`Town` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	`Postcode` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	`CompanyNo` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	`VatNo` TEXT NULL DEFAULT NULL COLLATE 'utf8mb4_0900_ai_ci',
	PRIMARY KEY (`id`) USING BTREE

![image](https://user-images.githubusercontent.com/75368133/169499105-2a811bf1-b742-43af-93bd-838c4a60db0a.png)

**Reply by melohagan** [reply]
Does your form definitely have the Customer field name set, e.g. click the dropdown and select 'Customer' similar to: 
![Screenshot 2022-05-20 at 10 41 37](https://user-images.githubusercontent.com/101575380/169501217-bee5ac5e-c5c7-463e-94d5-44c4dfec1c76.png)

**Reply by PaulMcF1987** [reply]
![image](https://user-images.githubusercontent.com/75368133/169502348-4ce72149-b538-45a9-aa83-33f8d4684ceb.png)

Ignore the key in the top box, thats my password manager lol

**Reply by melohagan** [reply]
I think I'll need to see the app. Are you willing to share the app export?

**Reply by PaulMcF1987** [reply]
@melohagan I can either share the app export or I can invite you as a user?

**Reply by melohagan** [reply]
Yes that works. Please invite the user: support@budibase.com

**Reply by PaulMcF1987** [reply]
@melohagan I have sent the invite and given development and admin permissions.

**Reply by melohagan** [reply]
That was a link to localhost. If you're only hosting locally atm, send the app export to that email address.

**Reply by PaulMcF1987** [reply]
I have updated it and resent the invite, it should be an IP Address, but I hadnt changed this in the settings.
I have also emailed and export to the support email address

**Reply by melohagan** [reply]
Thank you. Could you also unlock the app so I can have a look please

**Reply by melohagan** [reply]
So I've had a look - there's no design section? I think this goes back to this answer here: https://github.com/Budibase/budibase/discussions/5828#discussioncomment-2730296

You'll need a form that will pass through the bindings to your query. You can set default binding values in your query for the sake of testing it.

**Reply by PaulMcF1987** [reply]
@melohagan 
There is a design section, I have set all designs at present to `Power` access.
As I plan to open the app up to others with less access but want to ensure that everything works as planned with the admin user access first.

**Reply by melohagan** [reply]
Thanks, I was able to save this query:
![Screenshot 2022-05-20 at 12 01 55](https://user-images.githubusercontent.com/101575380/169514737-9478d16e-1100-4950-933a-9129c693cbb6.png)
 
Although I'm not sure why you need this if you have the Save Row action: 
![Screenshot 2022-05-20 at 12 02 43](https://user-images.githubusercontent.com/101575380/169514846-36e3fee6-e630-46e7-9d28-d924c2aec714.png)

**Reply by PaulMcF1987** [reply]
The save row action is there because I duplicated the `Edit form`  and it has all of the same details (the section in question was `recordjob/:id)
I was going to remove the save row tab and run the insert query instead.

I had been using `{ Form.Fields.JobTitle }}, {{ Form.Fields.Customer }}, {{ Form.Fields.BookedBy }}...`

**Reply by PaulMcF1987** [reply]
@melohagan 
![image](https://user-images.githubusercontent.com/75368133/169516618-3a99d51c-59ef-4f2f-89e8-7fe79a41cba3.png)

That is what I get with the full query, I have removed the `Form.Fields.` from the handlebars

**Reply by PaulMcF1987** [reply]
I have added a binding to all the entries
![image](https://user-images.githubusercontent.com/75368133/169523244-324e3241-66c9-40d8-aba2-c286b9ac2b8e.png)

And within the button action
![image](https://user-images.githubusercontent.com/75368133/169523366-f728c962-7718-4068-b7c0-e25513fe5b0b.png)

Tried it out and got
![image](https://user-images.githubusercontent.com/75368133/169523413-87870efb-cb73-401e-b080-c3982c40c644.png)


The query as is should work... I have used the exact same query in other apps and never had an issue

**Reply by melohagan** [reply]
Ah I think if the customer column was added manually as a relationship type, then it must not add it to the underlying MySQL table. 
So the insert is probably failing because the 'Customer' column does not exist in the MySQL table.

**Reply by PaulMcF1987** [reply]
@melohagan 
The query was taken from a previous app however the table and columns were created via Budibase
All of the fields in the database were added using
![image](https://user-images.githubusercontent.com/75368133/169529949-c9c91a5e-438c-49c6-90bb-a60924f3d3ca.png)
New Table
then 
![image](https://user-images.githubusercontent.com/75368133/169530030-cc045322-dc9e-4bf6-b211-9ced37df7287.png)
create column and setting up as required, Customer was set up as a relationship between itself and `Account`

The binding is recognised in the Button `Define Actions` panel
![image](https://user-images.githubusercontent.com/75368133/169532691-f9687690-61e8-43d5-a123-996e318301e8.png)

**Reply by PaulMcF1987** [reply]
@melohagan 
Any thoughts on the SQL Queries.

I am having a similar issue with a row that is saved as `formula`

I have used a formula to calculate a total cost for each row.
I then intended to use the totals column in a graph along with the month so the I could get a visual for the totals for each month

SELECT SUM(CustTotal) As Total, MONTH(CollectionDate) FROM JobRecord
GROUP BY MONTH(CollectionDate)

However, when I run the query I get the error `Query Error: Unknown column 'CustTotal' in 'field list'`

If I cannot use a query to get the information I need, is there another way to do it?

**Reply by melohagan** [reply]
So the `Query Error: Unknown column` means that the column was not found in the SQL table.

This is because you cannot use Formula columns in your SQL queries, as they are client side and used only within Budibase as display values.

**Reply by PaulMcF1987** [reply]
@melohagan I had an idea this was the case.

Is there an alternative method...

either to display the graph as detailed above, or to have a text field that executes the formula, that I can then use in a query?

(Although I have tried using the formula in a form field and it appreas to do nothing)
